e2e/v2_curl_test.go

```diff
@@ -48,66 +49,121 @@ func testCurlPutGet(t *testing.T, cfg *etcdProcessClusterConfig) {
+		t.Fatalf("failed get with curl (%v)", err)
+	}
 	if cfg.clientTLS == clientTLSAndNonTLS {
-		if err := cURLPut(epc, "testKey", "foo", expectPut); err != nil {
-			t.Fatalf("failed put with curl (%v)", err)
-		}
-
-		if err := cURLGet(epc, "testKey", expectGet); err != nil {
+		if err := cURL(epc, cURLReq{method: "GET", endpoint: "/v2/keys/foo", expected: expectGet, isTLS: true}); err != nil {
 			t.Fatalf("failed get with curl (%v)", err)
 		}
-		if err := cURLGetUseTLS(epc, "testKey", expectGet); err != nil {
-			t.Fatalf("failed get with curl (%v)", err)
-		}
-	} else {
-		if err := cURLPut(epc, "testKey", "foo", expectPut); err != nil {
-			t.Fatalf("failed put with curl (%v)", err)
+	}
+}
+
+func TestV2CurlIssue5182(t *testing.T) {
+	defer testutil.AfterTest(t)
+
+	epc := setupEtcdctlTest(t, &configNoTLS, false)
+	defer func() {
+		if err := epc.Close(); err != nil {
+			t.Fatalf("error closing etcd processes (%v)", err)
 		}
+	}()
 
-		if err := cURLGet(epc, "testKey", expectGet); err != nil {
-			t.Fatalf("failed get with curl (%v)", err)
+	expectPut := `{"action":"set","node":{"key":"/foo","value":"bar","`
+	if err := cURL(epc, cURLReq{method: "PUT", endpoint: "/v2/keys/foo", value: "bar", expected: expectPut}); err != nil {
+		t.Fatal(err)
+	}
+
+	expectUserAdd := `{"user":"foo","roles":null}`
+	if err := cURL(epc, cURLReq{method: "PUT", endpoint: "/v2/auth/users/foo", value: `{"user":"foo", "password":"pass"}`, expected: expectUserAdd, isJSON: true}); err != nil {
+		t.Fatal(err)
+	}
+	expectRoleAdd := `{"role":"foo","permissions":{"kv":{"read":["/foo/*"],"write":null}}`
+	if err := cURL(epc, cURLReq{method: "PUT", endpoint: "/v2/auth/roles/foo", value: `{"role":"foo", "permissions": {"kv": {"read": ["/foo/*"]}}}`, expected: expectRoleAdd, isJSON: true}); err != nil {
+		t.Fatal(err)
+	}
+	expectUserUpdate := `{"user":"foo","roles":["foo"]}`
+	if err := cURL(epc, cURLReq{method: "PUT", endpoint: "/v2/auth/users/foo", value: `{"user": "foo", "grant": ["foo"]}`, expected: expectUserUpdate, isJSON: true}); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := etcdctlUserAdd(epc, "root", "a"); err != nil {
+		t.Fatal(err)
+	}
+	if err := etcdctlAuthEnable(epc); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := cURL(epc, cURLReq{method: "GET", endpoint: "/v2/keys/foo", username: "root", password: "a", expected: "bar"}); err != nil {
+		t.Fatal(err)
+	}
+	if err := cURL(epc, cURLReq{method: "GET", endpoint: "/v2/keys/foo/", username: "foo", password: "pass", expected: "bar"}); err != nil {
+		t.Fatal(err)
+	}
+	if err := cURL(epc, cURLReq{method: "GET", endpoint: "/v2/keys/foo/", username: "foo", password: "", expected: "bar"}); err != nil {
+		if !strings.Contains(err.Error(), `The request requires user authentication`) {
+			t.Fatalf("expected 'The request requires user authentication' error, got %v", err)
 		}
+	} else {
+		t.Fatalf("expected 'The request requires user authentication' error")
 	}
 }
 
+type cURLReq struct {
+	method   string
```

Comment:
can there be separate curlGet and curlPut commands like before? setting this is really wordy